plaso/parsers/mac_keychain.py

```diff
@@ -183,298 +142,719 @@ class KeychainParser(interface.FileObjectParser):
+    Args:
+      attribute_values_data (bytes): attribute values data.
+      record_offset (int): offset of the record relative to the start of
+          the file.
+      attribute_values_data_offset (int): offset of the attribute values data
+          relative to the start of the record.
+      attribute_value_offset (int): offset of the attribute relative to
+          the start of the record.
 
-      creation_time = dfdatetime_time_elements.TimeElements()
-      try:
-        creation_time.CopyFromStringTuple(
-            time_elements_tuple=time_elements_tuple)
-      except ValueError:
-        creation_time = None
-        parser_mediator.ProduceExtractionError(
-            'invalid creation time value: {0!s}'.format(time_elements_tuple))
+    Returns:
+      str: string value or None if attribute value offset is not set.
+
+    Raises:
+      ParseError: if the attribute value cannot be read.
+    """
+    if attribute_value_offset == 0:
+      return None
 
-    last_modification_time = None
+    data_type_map = self._GetDataTypeMap('keychain_string')
 
-    structure_offset = record_offset + record.last_modification_time - 1
-    file_object.seek(structure_offset, os.SEEK_SET)
+    file_offset = (
+        record_offset + attribute_values_data_offset + attribute_value_offset)
+
+    attribute_value_offset -= attribute_values_data_offset + 1
+    attribute_value_data = attribute_values_data[attribute_value_offset:]
+
+    try:
+      string_attribute_value = self._ReadStructureFromByteStream(
+          attribute_value_data, file_offset, data_type_map)
+    except (ValueError, errors.ParseError) as exception:
+      raise errors.ParseError((
+          'Unable to map string attribute value data at offset: 0x{0:08x} '
+          'with error: {1!s}').format(file_offset, exception))
+
+    return string_attribute_value.string
+
+  def _ReadFileHeader(self, file_object):
+    """Reads the file header.
+
+    Args:
+      file_object (file): file-like object.
+
+    Returns:
+      keychain_file_header: file header.
+
+    Raises:
+      ParseError: if the file header cannot be read.
+    """
+    data_type_map = self._GetDataTypeMap('keychain_file_header')
+
+    file_header, _ = self._ReadStructureFromFileObject(
+        file_object, 0, data_type_map)
+
+    if file_header.signature != self._FILE_SIGNATURE:
+      raise errors.ParseError('Unsupported file signature.')
+
+    if (file_header.major_format_version != self._MAJOR_VERSION or
+        file_header.minor_format_version != self._MINOR_VERSION):
+      raise errors.ParseError('Unsupported format version: {0:s}.{1:s}'.format(
+          file_header.major_format_version, file_header.minor_format_version))
+
+    return file_header
+
+  def _ReadRecord(self, tables, file_object, record_offset, record_type):
+    """Reads the record.
+
+    Args:
+      tables (dict[str, KeychainDatabaseTable]): tables per name.
```

Comment:
Should be 'str' be 'int' here? On line 363, you use an int as a key into this dict.

Reply:
made some changes.